metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/validation/StellarZookeeperBasedValidator.java

```diff
@@ -0,0 +1,106 @@
+/*
+ *
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.apache.metron.stellar.common.utils.validation;
+
+import static org.apache.metron.stellar.common.shell.StellarShell.ERROR_PROMPT;
+
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.commons.lang.NullArgumentException;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.metron.stellar.common.StellarProcessor;
+import org.atteo.classindex.ClassIndex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class StellarZookeeperBasedValidator implements StellarValidator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final String FAILED_COMPILE = "Failed to compile";
+  private CuratorFramework client;
+
+  public StellarZookeeperBasedValidator(CuratorFramework client) throws NullArgumentException {
+    if (client == null) {
+      throw new NullArgumentException("client");
+    }
+    this.client = client;
+  }
+
+
+  @Override
+  public Iterable<ValidationResult> validate(Optional<LineWriter> writer) {
+    // discover all the StellarConfigurationProvider
+    Set<StellarConfigurationProvider> providerSet = new HashSet<>();
+
+    for (Class<?> c : ClassIndex.getSubclasses(StellarConfigurationProvider.class,
```

Comment:
If I understand this correctly, it seems that the `StellarConfigurationProvider` interface allows us to extend where configuration values get pulled in from.  In your current default implementation `ConfigurationProvider` you reach out to Zookeeper to pull in the config.

If I wanted to validate configuration located on a file system, I would just create a `FilesystemConfigurationProvider` implementation of this interface.

The decision as to whether I want to validate the config in Zookeeper, on the file system or both, needs to be user driven.  A user should make that decision based on how they call your new Stellar function.  

Based on your discovery logic here, just having a `FilesystemConfigurationProvider` on the classpath (or any other implementation) will cause the configuration in the file system to get validated.  We don't want that to happen.  We want the user to control that behavior.

So I don't think we really need this discovery logic, which nicely simplifies things.  I think we could just alter the `StellarValidater` interface to make this relationship simpler and more straight forward.  The `StellarConfigurationProvider` just gets passed in.

```
StellarValidator {
   ...
   Iterable<ValidationResult> validate(StellarConfigurationProvider provider);
   ...
}
```

Reply:
The idea is that not only can we not know the details of the classes that hold the rules, but also, that stellar may be hosted by other things than metron, that 'know' how to provide those configurations.  

The problem with this isn't the discovery per se, but in that it is not correct given it's purpose and the implementation

Reply:
I am not following your explanation of why we need the discovery logic.  Can you try to explain it again?  In the latest commits, I still see the discovery logic in StellarSimpleValidator.

Reply:
I will try to think of a different way to put it

Reply:
OK, I think I understand where you are coming from.  In the end, there will be more functions for validation.   Those which take in a passed in object, like CONFIG_GET() -> obj, EDIT -> obj, VALIDATE_CONFIG(obj).  Those functions will be user driven, ie.  __**Validate this thing I gave you__**.

These functions are __**Validate everything the system knows about**.  And are born out of the introduction of possible errors through language changes in already deployed stellar rules.

That is why we need discovery.  The "job" of this function at the moment is to find everything and make sure it is still OK.

Reply:
I think you're not understanding my point.   More specifically... What happens if another implementation gets added to the classpath?  

> Based on your discovery logic here, just having a FilesystemConfigurationProvider on the classpath (or any other implementation) will cause the configuration in the file system to get validated. We don't want that to happen. We want the user to control that behavior.

Reply:
They would automatically be picked up.

I'm confused.  This PR as stated is to help validation of stellar configurations, where ever they may be.  it is not for the user to hunt and peck and pick a configuration.  So that is kind of the point.

Although there have been comments about doing configuration validation with specific objects or files passed in, that is not the thrust of this pr, even if this pr enables it.

Reply:
But there is no need to discover those, right?  Why have the extra discovery logic when we don't need it?  

It is also more confusing from a user's perspective.  What I have on my class path decides what is going to get validated.  It is often very hard to control what gets added to a classpath.  That might lead to unexpected behavior.

If you just removed the discovery logic and instantiated the `StellarConfigurationProvider`'s that you want to use directly...

1. the functionality would work the same way, all the time, no matter what is on my class path.
1. the implementation would be simpler.

Reply:
You mean couple it to metron?  We don't want to do that anymore.

Reply:
Maybe that is where we are missing each other?

Reply:
> You mean couple it to metron? We don't want to do that anymore.

I don't see your justification.  Maybe another reviewer will understand this better.  As is, I am a +0 on this.

Reply:
https://issues.apache.org/jira/browse/METRON-989
https://issues.apache.org/jira/browse/METRON-876

@mattf-horton @cestella 

I have implemented this as to not increase the amount of tie-in between stellar and metron, and support future non-metron configuration sources. 

Thus, the sources of the configuration are discovered and not coupled.  I believe this is in the spirit if not the letter of the design discussions that we have had.

Can you take a look?

Reply:
It is true that Stellar should exist, as much as possible, independent from Metron; that was the aim of the effort to move stellar out of metron-common and into its own top level component in the project.  I'll look closer at this PR and the (rather long, but seemingly coherent...so congrats ;) comment thread and weigh in later.

Reply:
@cestella ping